pkg/package-server/provider/registry.go

```diff
@@ -131,43 +158,176 @@ func (p *RegistryProvider) syncCatalogSource(obj interface{}) (syncError error)
+	if _, syncError = p.sources.Add(key, address); syncError != nil {
+		logger.Warn("failed to create a new source")
+	}
+
+	return
+}
+
+func (p *RegistryProvider) syncSourceState(state registrygrpc.SourceState) {
+	key := state.Key
+	logger := logrus.WithFields(logrus.Fields{
+		"action": "sync source",
+		"source": key,
+		"state":  state.State,
+	})
+	logger.Debug("source state changed")
+
+	timeout, cancel := context.WithTimeout(context.Background(), cacheTimeout)
+	defer cancel()
+
+	var err error
+	switch state.State {
+	case connectivity.Ready:
+		var client *registryClient
+		client, err = p.registryClient(key)
+		if err == nil {
+			err = p.refreshCache(timeout, client)
 		}
+	case connectivity.TransientFailure, connectivity.Shutdown:
+		err = p.gcPackages(key, nil)
+	default:
+		logger.Debug("inert source state, skipping cache update")
+	}
 
-		logger.Info("grpc connection reset")
+	if err != nil {
+		logger.WithError(err).Warn("failed to update cache")
+	}
+}
+
+func (p *RegistryProvider) registryClient(key resolver.CatalogKey) (client *registryClient, err error) {
+	source := p.sources.Get(key)
+	if source == nil {
+		err = fmt.Errorf("missing source for catalog %s", key)
 		return
-	} else if ok {
-		// Address type grpc CatalogSource, drop the connection dial in to the new address
-		client.conn.Close()
 	}
 
-	logger.Info("attempting to add a new grpc connection")
-	conn, err := grpc.Dial(source.Address(), grpc.WithInsecure())
-	if err != nil {
-		logger.WithField("err", err.Error()).Errorf("could not connect to registry service")
-		syncError = err
+	conn := source.Conn
+	if conn == nil {
+		err = fmt.Errorf("missing grpc connection for source %s", key)
 		return
 	}
 
-	p.setClient(newRegistryClient(source, conn), key)
-	logger.Info("new grpc connection added")
+	var catsrc *operatorsv1alpha1.CatalogSource
+	catsrc, err = p.catsrcLister.CatalogSources(key.Namespace).Get(key.Name)
+	if err != nil {
+		return
+	}
 
+	client = newRegistryClient(catsrc, conn)
 	return
 }
 
+func (p *RegistryProvider) refreshCache(ctx context.Context, client *registryClient) error {
+	key, err := client.key()
+	if err != nil {
+		return err
+	}
+
+	logger := logrus.WithFields(logrus.Fields{
+		"action": "refresh cache",
+		"source": key,
+	})
+
+	stream, err := client.ListPackages(ctx, &api.ListPackageRequest{})
+	if err != nil {
+		logger.WithField("err", err.Error()).Warnf("error getting stream")
+		return nil
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+	added := map[string]struct{}{}
+	var mu sync.Mutex
+	for {
+		pkgName, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			logger.WithField("err", err.Error()).Warnf("error getting data")
+			break
+		}
+
+		g.Go(func() (pkgErr error) {
+			pkg, err := client.GetPackage(ctx, &api.GetPackageRequest{Name: pkgName.GetName()})
+			if err != nil {
+				logger.WithField("err", err.Error()).Warnf("eliding package: error getting package")
+				return
```

Comment:
In the anonymous function you defined a named-error return type `pkgErr` but then don't set `pkgErr` equal to the returned error before returning - does this function return the intended error or nil? Might be simpler to just return the error?

Reply:
Ah, I found I didn't want to return an error anywhere since this should be best effort, and didn't want to write `return nil` 4 times. l like the way errorgroups manage goroutines, but it may not be the best abstraction to use here. Any suggestions, using a waitgroup leaves something to be desired 🙁

Reply:
Oh I see what you mean...You're saying based off the `*Group.Go()` signature you need the anonymous function to to return an error (https://godoc.org/golang.org/x/sync/errgroup#Group.Go) but you don't actually want `*Group.Wait()` to return a non-nil error if one does come up. Since `the derived Context is canceled the first time a function passed to Go returns a non-nil error` and this is best effort we don't want that. 

Maybe just have `_ := g.Wait()` instead and comment that the error is being dropped intentionally as this is best effort to make it more clear for a reader? I see what you mean about the abstraction though

Reply:
@exdx So I took a look back at error group and found that some assumptions I made around how it handles context cancellation were wrong. I'll switch this to using `WaitGroup` instead and then we won't be returning an error from the anonymous function anyway. Should have this change up soon.